File: android/app/src/main/java/com/betterlise/app/ui/absences/AbsencesViewModel.kt
package com.betterlise.app.ui.absences

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.betterlise.app.data.api.AbsencesResponse
import com.betterlise.app.data.api.Endpoints
import com.betterlise.app.data.auth.SessionRepository
import com.betterlise.app.data.auth.SessionState
import com.betterlise.app.data.api.LiseHealth
import com.betterlise.app.data.cache.ResponseCache
import com.betterlise.app.data.health.LiseHealthMonitor
import com.betterlise.app.ui.loading.SyncState
import com.betterlise.app.ui.components.Loadable
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class AbsencesUiState(
    val isSignedIn: Boolean = false,
    val absences: Loadable<AbsencesResponse> = Loadable.Idle,
    val sync: SyncState = SyncState.Idle,
)

class AbsencesViewModel(
    private val session: SessionRepository,
    private val cache: ResponseCache,
    private val healthMonitor: LiseHealthMonitor,
    private val clock: () -> Long = System::currentTimeMillis,
) : ViewModel() {
    val health: StateFlow<LiseHealth?> = healthMonitor.health
    private val _state = MutableStateFlow(AbsencesUiState())
    val state: StateFlow<AbsencesUiState> = _state.asStateFlow()
    private var loadJob: Job? = null

    init {
        viewModelScope.launch {
            session.state.collect { sessionState ->
                val signedIn = sessionState is SessionState.SignedIn
                _state.value = AbsencesUiState(isSignedIn = signedIn)
                if (signedIn) refresh()
            }
        }
    }

    fun refresh() {
        loadJob?.cancel()
        loadJob = viewModelScope.launch {
            val cached = cache.load(CACHE_KEY, AbsencesResponse.serializer()) ?: _state.value.absences.value
            val startedAt = clock()
            val hasContent = cached != null
            _state.update { it.copy(absences = Loadable.Loading(cached), sync = SyncState.Syncing(startedAt, hasContent)) }
            launch { healthMonitor.refreshIfNeeded() }
            try {
                val response = session.send(Endpoints.absences())
                cache.save(CACHE_KEY, AbsencesResponse.serializer(), response)
                val finished = SyncState.Finished(startedAt, hasContent)
                _state.update { it.copy(absences = Loadable.Loaded(response), sync = finished) }
                completeSync(finished)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                _state.update { it.copy(absences = Loadable.Failed(e.message ?: "Erreur", cached), sync = SyncState.Idle) }
            }
        }
    }

    /** Shows "Terminé" briefly, like the web app, before revealing the content. */
    private fun completeSync(finished: SyncState.Finished) {
        viewModelScope.launch {
            delay(SyncState.COMPLETION_DISPLAY_MS)
            _state.update { if (it.sync == finished) it.copy(sync = SyncState.Idle) else it }
        }
    }

    private companion object {
        const val CACHE_KEY = "absences"
    }
}

File: ios/BetterLise/Features/Absences/AbsencesViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class AbsencesViewModel {
    private(set) var state: Loadable<AbsencesResponse> = .idle
    private(set) var syncState: SyncState = .idle

    let health: LiseHealthMonitor
    private let session: SessionStore
    private let cache: ResponseCache
    private let cacheKey = "absences"

    init(session: SessionStore, cache: ResponseCache, health: LiseHealthMonitor) {
        self.session = session
        self.cache = cache
        self.health = health
    }

    func load() async {
        let cached = cache.load(AbsencesResponse.self, key: cacheKey) ?? state.value
        state = .loading(cached: cached)
        let startedAt = Date()
        let hasContent = cached != nil
        syncState = .syncing(startedAt: startedAt, hasContent: hasContent)
        Task { await health.refreshIfNeeded() }

        do {
            let response = try await session.send(Endpoints.absences())
            cache.save(response, key: cacheKey)
            state = .loaded(response)
            completeSync(startedAt: startedAt, hasContent: hasContent)
        } catch {
            syncState = .idle
            state = .failed(message: error.localizedDescription, cached: cached)
        }
    }

    /// Shows the "Terminé" state briefly, like the web app, before revealing the content.
    private func completeSync(startedAt: Date, hasContent: Bool) {
        syncState = .finished(startedAt: startedAt, hasContent: hasContent)
        Task {
            try? await Task.sleep(for: SyncState.completionDisplayDuration)
            if syncState == .finished(startedAt: startedAt, hasContent: hasContent) {
                syncState = .idle
            }
        }
    }

    func reset() {
        state = .idle
        syncState = .idle
    }
}
